src/ddmd/root/filename.d

```diff
@@ -721,3 +731,132 @@ nothrow:
+            }
+        }
+
+        // different conventions for CreateDirectory and mkdir
+        return createRet == 0 ? 1 : 0;
+    }
+
+    // Converts a path to one suitable to be passed to Win32 API
+    // functions that can deal with paths longer than 248
+    // characters then calls the supplied function on it.
+    // For more information:
+    // https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx
+    package auto extendedPathThen(alias F)(const(char*) path)
+    {
+        wchar[1024] wpathBuf;
+        // Since the unicode Win32 APIs use UTF16, we need to convert to UTF16
+        const wpath = path.toWStringz(wpathBuf);
+
+        // GetFullPathNameW expects a sized buffer to store the result in. Since we don't
+        // know how larget it has to be, we pass in null and get the needed buffer length
+        // as the return code.
+        const pathLength = GetFullPathNameW(&wpath[0],
+                                            0 /*length8*/,
+                                            null /*output buffer*/,
+                                            null /*filePartBuffer*/);
+        if (pathLength == 0)
+        {
+            return F(""w);
+        }
+
+        // wpath is the UTF16 version of path, but to be able to use
+        // extended paths, we need to prefix with `\\?\` and the absolute
+        // path.
+        static immutable prefix = `\\?\`w;
+
+        // +1 for the null terminator
+        const bufferLength = pathLength + prefix.length + 1;
+
+        wchar[1024] absBuf;
+        auto absPath = bufferLength > absBuf.length ? new wchar[bufferLength] : absBuf[];
+
+        absPath[0 .. prefix.length] = prefix[];
+
+        const absPathRet = GetFullPathNameW(&wpath[0],
+                                            absPath.length - prefix.length,
+                                            &absPath[prefix.length],
+                                            null /*filePartBuffer*/);
+
+        if (absPathRet == 0 || absPathRet > absPath.length - prefix.length)
+        {
+            return F(""w);
+        }
+
+        auto extendedPath = absPath[0 .. absPathRet];
+        return F(extendedPath);
+    }
+
+    // Converts a null-terminated string to an array of wchar that's null
+    // terminated so it can be passed to Win32 APIs.
+    // buf is passed as a scratch space to store the result. If more memory
+    // is needed then toWstringz allocates on the GC heap instead.
+    private wchar[] toWStringz(const(char*) str, wchar[] buf = []) nothrow
+    {
+        import core.stdc.string: strlen;
+
+        // cache this for efficiency
+        const strLength = strlen(str) + 1;
+        // first find out how long the buffer must be to store the result
+        const length = MultiByteToWideChar(0 /*codepage*/, 0 /*flags*/, str, strLength, null, 0);
+        if (!length) return null;
+
+        auto ret = length > buf.length ? new wchar[length] : buf;
```

Comment:
The `new` is a memory leak in dmd. Take a look at https://github.com/dlang/phobos/blob/master/std/internal/cstring.d#L122 where malloc is used for the temporary buffer, cleaned up with free().

Reply:
I thought that leaking was the memory strategy for dmd in order for it to be fast and that at some point the GC would be turned on. The code you linked is from Phobos, so do you mean I should create a similar struct for dmd or just to use `free` to clean up?